File: java/141.环形链表.java
/*
 * @Author: your name
 * @Date: 2019-12-12 14:14:55
 * @LastEditTime: 2019-12-12 14:21:13
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \leetcode\141.环形链表.java
 */
/*
 * @lc app=leetcode.cn id=141 lang=java
 *
 * [141] 环形链表 快慢指针，想想一下，在一个环形的跑道上，
 * 两个运动员一快一慢，最终快的那个运动员会追上慢的那个
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head== null || head.next == null){
            return false;
        }
        ListNode slow  = head;
        ListNode fast = head.next;
        
        while(slow != fast){
            if(fast == null || fast.next==null){
                return false;
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        return true;
    }
}
// @lc code=end

File: python/141.环形链表.py
'''
@Author: your name
@Date: 2019-12-11 19:24:21
@LastEditTime: 2019-12-12 14:14:30
@LastEditors: Please set LastEditors
@Description: In User Settings Edit
@FilePath: \leetcode\141.环形链表.py
'''
#
# @lc app=leetcode.cn id=141 lang=python3
#
# [141] 环形链表
#想象一下，两名运动员以不同的速度在环形赛道上跑步会发生什么？

# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if(head== None or head.next == None):
            return False
        slow = head
        fast = head.next
        
        while slow != fast:
            if fast==None or fast.next == None:
                return False
            slow = slow.next
            fast = fast.next.next
        return True

a = Solution()
root = ListNode(3)
a.hasCycle(root)
# @lc code=end

